chain/client/src/state_request_actor.rs

```diff
@@ -0,0 +1,237 @@
+use std::collections::VecDeque;
+use std::sync::Arc;
+
+use near_async::messaging::{Actor, Handler};
+use near_async::time::{Clock, Duration, Instant};
+use near_chain::state_sync::ChainStateSyncAdapter;
+use near_chain::types::RuntimeAdapter;
+use near_epoch_manager::EpochManagerAdapter;
+use near_network::client::{StateRequestHeader, StateRequestPart, StateResponse};
+use near_network::types::{StateResponseInfo, StateResponseInfoV2};
+use near_performance_metrics_macros::perf;
+use near_primitives::hash::CryptoHash;
+use near_primitives::state_sync::{
+    ShardStateSyncResponse, ShardStateSyncResponseHeader, ShardStateSyncResponseV3,
+};
+use near_store::adapter::chain_store::ChainStoreAdapter;
+use parking_lot::Mutex;
+use tracing::{error, info, warn};
+
+use crate::metrics;
+
+pub struct StateRequestActor {
+    clock: Clock,
+    state_sync_adapter: ChainStateSyncAdapter,
+    chain_store: ChainStoreAdapter,
+    genesis_hash: CryptoHash,
+    throttle_period: Duration,
+    num_state_requests_per_throttle_period: usize,
+    state_request_cache: Arc<Mutex<VecDeque<Instant>>>,
```

Comment:
Preexisting, but naming nit: this probably isn't a cache if it only holds instants and not state requests.